File: frontend/src/optical/kinds/laser-source/physics.ts
/**
 * Laser source PhysicsOp -- v3.
 *
 * Op name: `emit_laser_source`
 * Kind:    `laser_source`
 *
 * A laser source is a scene emitter, not a passive element. Its Asset3D
 * should expose one optical face, normally `out`; the face normal is the
 * emitted chief-ray direction in the asset body frame.
 */

import {
  type BeamRay,
  type Vec3,
  normalize,
} from "../../beam-ray";
import { type Complex, qAtWaist } from "../../fiber/gaussian";
import {
  type PhysicsOp,
  type PhysicsOpContext,
  registerKind,
} from "../../registry";

const ZERO_C: Complex = { re: 0, im: 0 };
const ONE_C: Complex = { re: 1, im: 0 };

function finiteNumber(v: unknown): number | undefined {
  return typeof v === "number" && Number.isFinite(v) ? v : undefined;
}

function positiveFiniteNumber(v: unknown): number | undefined {
  const n = finiteNumber(v);
  return n !== undefined && n > 0 ? n : undefined;
}

function readPath(obj: unknown, path: string[]): unknown {
  let cur = obj;
  for (const key of path) {
    if (!cur || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[key];
  }
  return cur;
}

function firstPositive(ctx: PhysicsOpContext, paths: string[][], fallback: number): number {
  for (const path of paths) {
    const d = positiveFiniteNumber(readPath(ctx.dynamic, path));
    if (d !== undefined) return d;
  }
  for (const path of paths) {
    const p = positiveFiniteNumber(readPath(ctx.params, path));
    if (p !== undefined) return p;
  }
  return fallback;
}

function firstFinite(ctx: PhysicsOpContext, paths: string[][], fallback: number): number {
  for (const path of paths) {
    const d = finiteNumber(readPath(ctx.dynamic, path));
    if (d !== undefined) return d;
  }
  for (const path of paths) {
    const p = finiteNumber(readPath(ctx.params, path));
    if (p !== undefined) return p;
  }
  return fallback;
}

function complexFromUnknown(v: unknown): Complex | undefined {
  if (!v || typeof v !== "object") return undefined;
  const r = v as Record<string, unknown>;
  const re = finiteNumber(r.re);
  const im = finiteNumber(r.im) ?? 0;
  return re !== undefined ? { re, im } : undefined;
}

function jonesFromUnknown(v: unknown): [Complex, Complex] | undefined {
  if (Array.isArray(v) && v.length >= 2) {
    const a = complexFromUnknown(v[0]);
    const b = complexFromUnknown(v[1]);
    if (a && b) return [a, b];
  }
  if (v && typeof v === "object") {
    const r = v as Record<string, unknown>;
    const nested = jonesFromUnknown(r.jones);
    if (nested) return nested;
    const exRe = finiteNumber(r.exRe);
    const eyRe = finiteNumber(r.eyRe);
    if (exRe !== undefined && eyRe !== undefined) {
      return [
        { re: exRe, im: finiteNumber(r.exIm) ?? 0 },
        { re: eyRe, im: finiteNumber(r.eyIm) ?? 0 },
      ];
    }
  }
  return undefined;
}

function readJones(ctx: PhysicsOpContext): [Complex, Complex] {
  const candidates = [
    ctx.dynamic?.jones,
    ctx.dynamic?.polarization,
    readPath(ctx.dynamic, ["beam", "jones"]),
    readPath(ctx.dynamic, ["beam", "polarization"]),
    ctx.params.jones,
    ctx.params.polarization,
  ];
  for (const c of candidates) {
    const parsed = jonesFromUnknown(c);
    if (parsed) return parsed;
  }
  return [ONE_C, ZERO_C];
}

function qFromMode(
  waistUm: number,
  waistZOffsetMm: number,
  wavelengthNm: number,
): Complex {
  const waistMm = waistUm / 1000;
  const lambdaMm = wavelengthNm * 1e-6;
  const q = qAtWaist(waistMm, lambdaMm);
  return { re: -waistZOffsetMm, im: q.im };
}

export function emitLaserSourceRay(ctx: PhysicsOpContext): BeamRay {
  const wavelengthNm = firstPositive(ctx, [
    ["centerWavelengthNm"],
    ["spectrum", "centerWavelengthNm"],
    ["beam", "centerWavelengthNm"],
    ["beam", "spectrum", "centerWavelengthNm"],
  ], 780.241);
  const powerMw = firstFinite(ctx, [
    ["laserPowerMw"],
    ["powerMw"],
    ["nominalPowerMw"],
    ["beam", "laserPowerMw"],
    ["beam", "powerMw"],
    ["beam", "nominalPowerMw"],
  ], 1.0);

  const waistXUm = firstPositive(ctx, [
    ["spatialModeX", "waistUm"],
    ["spatialEnvelope", "waistXUm"],
    ["spatialEnvelope", "waistUm"],
    ["spatialEnvelope", "transverseProfile", "x", "waistRadiusUm"],
    ["waistUm"],
    ["beam", "spatialModeX", "waistUm"],
    ["beam", "spatialEnvelope", "waistXUm"],
    ["beam", "spatialEnvelope", "waistUm"],
    ["beam", "spatialEnvelope", "transverseProfile", "x", "waistRadiusUm"],
    ["beam", "waistUm"],
  ], 250);
  const waistYUm = firstPositive(ctx, [
    ["spatialModeY", "waistUm"],
    ["spatialEnvelope", "waistYUm"],
    ["spatialEnvelope", "waistUm"],
    ["spatialEnvelope", "transverseProfile", "y", "waistRadiusUm"],
    ["waistUm"],
    ["beam", "spatialModeY", "waistUm"],
    ["beam", "spatialEnvelope", "waistYUm"],
    ["beam", "spatialEnvelope", "waistUm"],
    ["beam", "spatialEnvelope", "transverseProfile", "y", "waistRadiusUm"],
    ["beam", "waistUm"],
  ], waistXUm);
  const waistXOffsetMm = firstFinite(ctx, [
    ["spatialModeX", "waistZOffsetMm"],
    ["spatialEnvelope", "waistXOffsetMm"],
    ["spatialEnvelope", "waistZOffsetMm"],
    ["spatialEnvelope", "propagation", "x", "waistZOffsetMm"],
    ["beam", "spatialModeX", "waistZOffsetMm"],
    ["beam", "spatialEnvelope", "waistXOffsetMm"],
    ["beam", "spatialEnvelope", "waistZOffsetMm"],
    ["beam", "spatialEnvelope", "propagation", "x", "waistZOffsetMm"],
  ], 0);
  const waistYOffsetMm = firstFinite(ctx, [
    ["spatialModeY", "waistZOffsetMm"],
    ["spatialEnvelope", "waistYOffsetMm"],
    ["spatialEnvelope", "waistZOffsetMm"],
    ["spatialEnvelope", "propagation", "y", "waistZOffsetMm"],
    ["beam", "spatialModeY", "waistZOffsetMm"],
    ["beam", "spatialEnvelope", "waistYOffsetMm"],
    ["beam", "spatialEnvelope", "waistZOffsetMm"],
    ["beam", "spatialEnvelope", "propagation", "y", "waistZOffsetMm"],
  ], waistXOffsetMm);

  const direction: Vec3 = normalize(ctx.faceOut.normalBodyLocal ?? { x: 0, y: 0, z: 1 });
  return {
    origin: ctx.faceOut.positionMmBodyLocal,
    direction,
    qx: qFromMode(waistXUm, waistXOffsetMm, wavelengthNm),
    qy: qFromMode(waistYUm, waistYOffsetMm, wavelengthNm),
    wavelengthNm,
    powerMw: Math.max(0, powerMw),
    jones: readJones(ctx),
    pathLengthMm: 0,
    phaseAccumRad: 0,
  };
}

export const emitLaserSourceOp: PhysicsOp = (
  _rayIn: BeamRay,
  ctx: PhysicsOpContext,
): BeamRay[] => [emitLaserSourceRay(ctx)];

registerKind("laser_source", {
  ops: {
    emit_laser_source: emitLaserSourceOp,
  },
  needsAperture: false,
});

File: backend/app/optical/kinds/laser_source/physics.py
"""Laser source v3 PhysicsOp (Python mirror of frontend kind).

Op name: emit_laser_source
Kind: laser_source

A laser source is a scene emitter, not a passive element. Its Asset3D should
expose one optical face, normally ``out``; the face normal is the emitted
chief-ray direction in asset body coordinates.
"""

from __future__ import annotations

import math
from typing import Optional

from app.optical.beam_ray import BeamRay, Vec3, q_at_waist
from app.optical.registry import (
    KindEntry,
    PhysicsOpContext,
    register_kind,
)


def _finite_number(value) -> Optional[float]:
    if isinstance(value, (int, float)) and math.isfinite(float(value)):
        return float(value)
    return None


def _positive_finite_number(value) -> Optional[float]:
    n = _finite_number(value)
    if n is not None and n > 0:
        return n
    return None


def _read_path(obj, path: list[str]):
    cur = obj
    for key in path:
        if not isinstance(cur, dict):
            return None
        cur = cur.get(key)
    return cur


def _first_positive(ctx: PhysicsOpContext, paths: list[list[str]], fallback: float) -> float:
    dynamic = ctx.dynamic or {}
    for path in paths:
        d = _positive_finite_number(_read_path(dynamic, path))
        if d is not None:
            return d
    for path in paths:
        p = _positive_finite_number(_read_path(ctx.params, path))
        if p is not None:
            return p
    return fallback


def _first_finite(ctx: PhysicsOpContext, paths: list[list[str]], fallback: float) -> float:
    dynamic = ctx.dynamic or {}
    for path in paths:
        d = _finite_number(_read_path(dynamic, path))
        if d is not None:
            return d
    for path in paths:
        p = _finite_number(_read_path(ctx.params, path))
        if p is not None:
            return p
    return fallback


def _complex_from_unknown(value) -> Optional[complex]:
    if not isinstance(value, dict):
        return None
    re = _finite_number(value.get("re"))
    im = _finite_number(value.get("im")) or 0.0
    if re is None:
        return None
    return complex(re, im)


def _jones_from_unknown(value) -> Optional[tuple[complex, complex]]:
    if isinstance(value, list) and len(value) >= 2:
        a = _complex_from_unknown(value[0])
        b = _complex_from_unknown(value[1])
        if a is not None and b is not None:
            return (a, b)
    if isinstance(value, dict):
        nested = _jones_from_unknown(value.get("jones"))
        if nested is not None:
            return nested
        ex_re = _finite_number(value.get("exRe"))
        ey_re = _finite_number(value.get("eyRe"))
        if ex_re is not None and ey_re is not None:
            return (
                complex(ex_re, _finite_number(value.get("exIm")) or 0.0),
                complex(ey_re, _finite_number(value.get("eyIm")) or 0.0),
            )
    return None


def _read_jones(ctx: PhysicsOpContext) -> tuple[complex, complex]:
    dynamic = ctx.dynamic or {}
    for candidate in (
        dynamic.get("jones"),
        dynamic.get("polarization"),
        _read_path(dynamic, ["beam", "jones"]),
        _read_path(dynamic, ["beam", "polarization"]),
        ctx.params.get("jones"),
        ctx.params.get("polarization"),
    ):
        parsed = _jones_from_unknown(candidate)
        if parsed is not None:
            return parsed
    return (complex(1, 0), complex(0, 0))


def _q_from_mode(waist_um: float, waist_z_offset_mm: float, wavelength_nm: float) -> complex:
    waist_mm = waist_um / 1000.0
    lambda_mm = wavelength_nm * 1e-6
    q = q_at_waist(waist_mm, lambda_mm)
    return complex(-waist_z_offset_mm, q.imag)


def emit_laser_source_ray(ctx: PhysicsOpContext) -> BeamRay:
    wavelength_nm = _first_positive(ctx, [
        ["centerWavelengthNm"],
        ["spectrum", "centerWavelengthNm"],
        ["beam", "centerWavelengthNm"],
        ["beam", "spectrum", "centerWavelengthNm"],
    ], 780.241)
    power_mw = _first_finite(ctx, [
        ["laserPowerMw"],
        ["powerMw"],
        ["nominalPowerMw"],
        ["beam", "laserPowerMw"],
        ["beam", "powerMw"],
        ["beam", "nominalPowerMw"],
    ], 1.0)

    waist_x_um = _first_positive(ctx, [
        ["spatialModeX", "waistUm"],
        ["spatialEnvelope", "waistXUm"],
        ["spatialEnvelope", "waistUm"],
        ["spatialEnvelope", "transverseProfile", "x", "waistRadiusUm"],
        ["waistUm"],
        ["beam", "spatialModeX", "waistUm"],
        ["beam", "spatialEnvelope", "waistXUm"],
        ["beam", "spatialEnvelope", "waistUm"],
        ["beam", "spatialEnvelope", "transverseProfile", "x", "waistRadiusUm"],
        ["beam", "waistUm"],
    ], 250.0)
    waist_y_um = _first_positive(ctx, [
        ["spatialModeY", "waistUm"],
        ["spatialEnvelope", "waistYUm"],
        ["spatialEnvelope", "waistUm"],
        ["spatialEnvelope", "transverseProfile", "y", "waistRadiusUm"],
        ["waistUm"],
        ["beam", "spatialModeY", "waistUm"],
        ["beam", "spatialEnvelope", "waistYUm"],
        ["beam", "spatialEnvelope", "waistUm"],
        ["beam", "spatialEnvelope", "transverseProfile", "y", "waistRadiusUm"],
        ["beam", "waistUm"],
    ], waist_x_um)
    waist_x_offset_mm = _first_finite(ctx, [
        ["spatialModeX", "waistZOffsetMm"],
        ["spatialEnvelope", "waistXOffsetMm"],
        ["spatialEnvelope", "waistZOffsetMm"],
        ["spatialEnvelope", "propagation", "x", "waistZOffsetMm"],
        ["beam", "spatialModeX", "waistZOffsetMm"],
        ["beam", "spatialEnvelope", "waistXOffsetMm"],
        ["beam", "spatialEnvelope", "waistZOffsetMm"],
        ["beam", "spatialEnvelope", "propagation", "x", "waistZOffsetMm"],
    ], 0.0)
    waist_y_offset_mm = _first_finite(ctx, [
        ["spatialModeY", "waistZOffsetMm"],
        ["spatialEnvelope", "waistYOffsetMm"],
        ["spatialEnvelope", "waistZOffsetMm"],
        ["spatialEnvelope", "propagation", "y", "waistZOffsetMm"],
        ["beam", "spatialModeY", "waistZOffsetMm"],
        ["beam", "spatialEnvelope", "waistYOffsetMm"],
        ["beam", "spatialEnvelope", "waistZOffsetMm"],
        ["beam", "spatialEnvelope", "propagation", "y", "waistZOffsetMm"],
    ], waist_x_offset_mm)

    direction = (ctx.face_out.normal_body_local or Vec3(0, 0, 1)).normalized()
    return BeamRay(
        origin=ctx.face_out.position_mm_body_local,
        direction=direction,
        qx=_q_from_mode(waist_x_um, waist_x_offset_mm, wavelength_nm),
        qy=_q_from_mode(waist_y_um, waist_y_offset_mm, wavelength_nm),
        wavelength_nm=wavelength_nm,
        power_mw=max(0.0, power_mw),
        jones=_read_jones(ctx),
    )


def emit_laser_source_op(_ray_in: BeamRay, ctx: PhysicsOpContext) -> list[BeamRay]:
    return [emit_laser_source_ray(ctx)]


register_kind("laser_source", KindEntry(
    ops={"emit_laser_source": emit_laser_source_op},
    needs_aperture=False,
))
